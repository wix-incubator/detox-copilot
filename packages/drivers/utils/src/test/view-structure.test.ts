import {
  TestContext,
  setupTestEnvironment,
  teardownTestEnvironment,
} from "./setup";

describe("View Structure Extraction", () => {
  let testContext: TestContext;

  beforeAll(async () => {
    testContext = await setupTestEnvironment();
  }, 30000);

  afterAll(async () => {
    await teardownTestEnvironment(testContext);
  });

  it("should generate clean DOM structure with only relevant elements", async () => {
    const structure = await testContext.page.evaluate(() => {
      window.driverUtils.markImportantElements();
      return window.driverUtils.extractCleanViewStructure();
    });

    expect(structure).toMatchSnapshot("clean-dom-structure");
  });

  it("should preserve allowed attributes and remove others", async () => {
    const attributesCheck = await testContext.page.evaluate(() => {
      // Add some elements with various attributes
      const div = document.createElement("div");
      div.innerHTML = `
        <button class="btn" style="color: red" data-test="value" onclick="alert()">Test</button>
        <a href="#link" target="_blank" rel="noopener" style="color: blue">Link</a>
        <img src="test.jpg" alt="Test" width="100" height="100">
      `;
      document.body.appendChild(div);

      window.driverUtils.markImportantElements();
      const structure = window.driverUtils.extractCleanViewStructure();
      document.body.removeChild(div);

      return structure;
    });

    expect(attributesCheck).toMatchSnapshot("attribute-filtering");
  });

  it("should preserve element hierarchy and relationships", async () => {
    const structure = await testContext.page.evaluate(() => {
      // Add complex nested structure
      const complex = document.createElement("div");
      complex.innerHTML = `
        <nav data-testid="main-nav">
          <ul>
            <li>
              <a href="#1">Link 1</a>
              <ul>
                <li><a href="#1.1">Sublink 1.1</a></li>
                <li><a href="#1.2">Sublink 1.2</a></li>
              </ul>
            </li>
          </ul>
        </nav>
      `;
      document.body.appendChild(complex);

      window.driverUtils.markImportantElements();
      const result = window.driverUtils.extractCleanViewStructure();
      document.body.removeChild(complex);

      return result;
    });

    // Verify semantic structure is preserved
    expect(structure).toMatch(
      /aria-pilot-category="semantic".*aria-pilot-category="list"/s,
    );

    // Verify proper nesting and closing tags
    expect(structure).toMatch(/<nav[^>]*>.*<\/nav>/s);
    expect(structure).toMatch(/<ul[^>]*>.*<\/ul>/s);

    // Verify categories are assigned correctly
    expect(structure).toMatch(/aria-pilot-category="semantic"[^>]*>.*<nav/s);
    expect(structure).toMatch(/aria-pilot-category="list"[^>]*>.*<\/ul>/s);
  });
});
