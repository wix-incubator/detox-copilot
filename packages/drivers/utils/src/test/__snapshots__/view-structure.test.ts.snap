// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`View Structure Extraction should generate clean DOM structure with only relevant elements: clean-dom-structure 1`] = `
"<html>
  <head>
  </head>
  <body>
    <header aria-pilot-category="semantic" aria-pilot-index="0" data-category="semantic" data-index="0">
      <h1 aria-pilot-category="header" aria-pilot-index="0" data-category="header" data-index="0">
      </h1>
      <nav aria-pilot-category="semantic" aria-pilot-index="1" data-category="semantic" data-index="1">
        <ul aria-pilot-category="list" aria-pilot-index="0" data-category="list" data-index="0">
        </ul>
      </nav>
    </header>
    <main aria-pilot-category="semantic" aria-pilot-index="2" data-category="semantic" data-index="2">
      <section aria-pilot-category="semantic" aria-pilot-index="3" data-category="semantic" data-index="3">
        <h2 aria-pilot-category="header" aria-pilot-index="1" data-category="header" data-index="1">
        </h2>
      </section>
      <section aria-pilot-category="semantic" aria-pilot-index="4" data-category="semantic" data-index="4">
        <h2 aria-pilot-category="header" aria-pilot-index="2" data-category="header" data-index="2">
        </h2>
        <ul aria-pilot-category="list" aria-pilot-index="1" data-category="list" data-index="1">
        </ul>
        <ol aria-pilot-category="list" aria-pilot-index="2" data-category="list" data-index="2">
        </ol>
      </section>
      <section aria-pilot-category="semantic" aria-pilot-index="5" data-category="semantic" data-index="5">
        <h2 aria-pilot-category="header" aria-pilot-index="3" data-category="header" data-index="3">
        </h2>
        <table aria-pilot-category="table" aria-pilot-index="0" data-category="table" data-index="0">
        </table>
      </section>
      <section aria-pilot-category="semantic" aria-pilot-index="6" data-category="semantic" data-index="6">
        <h2 aria-pilot-category="header" aria-pilot-index="4" data-category="header" data-index="4">
        </h2>
      </section>
    </main>
    <footer aria-pilot-category="semantic" aria-pilot-index="7" data-category="semantic" data-index="7">
    </footer>
  </body>
</html>
"
`;

exports[`View Structure Extraction should preserve allowed attributes and remove others: attribute-filtering 1`] = `
"<html>
  <head>
  </head>
  <body>
    <header aria-pilot-category="semantic" aria-pilot-index="0" data-category="semantic" data-index="0">
      <h1 aria-pilot-category="header" aria-pilot-index="0" data-category="header" data-index="0">
      </h1>
      <nav aria-pilot-category="semantic" aria-pilot-index="1" data-category="semantic" data-index="1">
        <ul aria-pilot-category="list" aria-pilot-index="0" data-category="list" data-index="0">
        </ul>
      </nav>
    </header>
    <main aria-pilot-category="semantic" aria-pilot-index="2" data-category="semantic" data-index="2">
      <section aria-pilot-category="semantic" aria-pilot-index="3" data-category="semantic" data-index="3">
        <h2 aria-pilot-category="header" aria-pilot-index="1" data-category="header" data-index="1">
        </h2>
      </section>
      <section aria-pilot-category="semantic" aria-pilot-index="4" data-category="semantic" data-index="4">
        <h2 aria-pilot-category="header" aria-pilot-index="2" data-category="header" data-index="2">
        </h2>
        <ul aria-pilot-category="list" aria-pilot-index="1" data-category="list" data-index="1">
        </ul>
        <ol aria-pilot-category="list" aria-pilot-index="2" data-category="list" data-index="2">
        </ol>
      </section>
      <section aria-pilot-category="semantic" aria-pilot-index="5" data-category="semantic" data-index="5">
        <h2 aria-pilot-category="header" aria-pilot-index="3" data-category="header" data-index="3">
        </h2>
        <table aria-pilot-category="table" aria-pilot-index="0" data-category="table" data-index="0">
        </table>
      </section>
      <section aria-pilot-category="semantic" aria-pilot-index="6" data-category="semantic" data-index="6">
        <h2 aria-pilot-category="header" aria-pilot-index="4" data-category="header" data-index="4">
        </h2>
      </section>
    </main>
    <footer aria-pilot-category="semantic" aria-pilot-index="7" data-category="semantic" data-index="7">
    </footer>
  </body>
</html>
"
`;
